tensorflow_model_optimization/python/core/sparsity_tf2/sparse_utils_test.py

```diff
@@ -0,0 +1,223 @@
+
+class SparseUtilsTest(test.TestCase, parameterized.TestCase):
+
+  def setUp(self):
+    super(SparseUtilsTest, self).setUp()
+    self.seeds = range(20)
+
+  @parameterized.parameters(
+    (-1.,), (1.5,)
+  )
+  def testInvalidRatioRaisesError(self, p):
+    with self.assertRaises(ValueError):
+      bernouilli = sparse_utils.Bernouilli(p)
+      permuteones = sparse_utils.PermuteOnes(p)
+
+  @parameterized.parameters(
+    (tf.int32,), (tf.float32,), (tf.int64,), (tf.float64,)
+  )
+  def testMatrixDtype(self, dtype):
+    shape = (3, 4)
+    seed = 0
+    ratio = 0.5
+    type_bern_checks = []
+    type_perm_checks = []
+    _dtype = [dtype]
+    matrix = tf.ones(shape)
+    sparse_bernouilli = sparse_utils.Bernouilli(ratio)(shape=matrix.shape, dtype=dtype, seed=seed)
+    type_bern_checks.append(sparse_bernouilli.dtype)
+    self.assertAllEqual(_dtype, type_bern_checks)
+    sparse_permuteones = sparse_utils.PermuteOnes(ratio)(shape=matrix.shape, dtype=dtype, seed=seed)
+    type_perm_checks.append(sparse_permuteones.dtype)
+    self.assertAllEqual(_dtype, type_perm_checks)
+    
+  @parameterized.parameters(
+    itertools.product((0.4,), NOSMALL_SHAPES)
+  )
+  def testScatteredInitialOnesPermuteOnes(self, ratio, shape):
+    # This test checks that the expected sparsity is reflected in the 
+    # average number of times each index in the matrix is a 1.
+    # note that ratio here is fixed because the shape / density
+    # would make a greater difference in this check for "scattered-ness".
+    # Shapes cannot be small (i.e. () or (1) ) as there are not enough
+    # samples to average over.
+    matrix = tf.ones(shape)
+    self.assertAllEqual(shape, matrix.shape)
+
+    seed = 0
+    sparse_matrices_perm = []
+    for _ in range(400):
+      matrix_sparse_permute = sparse_utils.PermuteOnes(ratio)(shape, seed=seed)
+      sparse_matrices_perm.append(matrix_sparse_permute)
+      seed += 1
+    average_initialization = tf.math.reduce_mean(sparse_matrices_perm, axis=0)
+    expected_initialization = tf.constant(ratio, shape=shape) # broadcasts
+    # expect that each entry in the matrix is "on"/1 in expectation, ratio number of times
+    self.assertAllClose(expected_initialization, average_initialization, rtol=231e-3)
+   
+  @parameterized.parameters(
+    itertools.product((0.4,), NOSMALL_SHAPES)
+  )
+  def testScatteredInitialOnesBernouilli(self, ratio, shape):
+    matrix = tf.ones(shape)
+    self.assertAllEqual(shape, matrix.shape)
+
+    seed = 0
+    sparse_matrices_bern = []
+    for _ in range(400):
+      matrix_sparse_bernouilli = sparse_utils.Bernouilli(ratio)(shape, seed=seed)
+      sparse_matrices_bern.append(matrix_sparse_bernouilli)
+      seed += 1
+    average_initialization = tf.math.reduce_mean(sparse_matrices_bern, axis=0)
+    expected_initialization = tf.constant(ratio, shape=shape)
+    self.assertAllClose(expected_initialization, average_initialization, rtol=204e-3)
```

Comment:
204e-3? Is that 2.04e-1? Which means we're asserting that it's w/in 20% of the value? If so can this be made a little more explicit (e.g. 2.1e-1 or 0.21 or something?)